File: app/src/main/java/com/jintao/vipmanager/database/DatabaseManager.java
package com.jintao.vipmanager.database;

import android.content.Context;
import com.jintao.vipmanager.BuildConfig;
import com.jintao.vipmanager.database.dao.DaoMaster;
import com.jintao.vipmanager.database.dao.DaoSession;
import com.jintao.vipmanager.utils.AppConfig;

import org.greenrobot.greendao.query.QueryBuilder;

/**
 * 所有数据库都在这儿打开关闭
 */
public class DatabaseManager {

    private Context context;

    //多线程中要被共享的使用volatile关键字修饰
    private volatile static DatabaseManager manager = new DatabaseManager();

    private static DbUpdateHelper mVipUserHelper;
    private static DaoMaster mVipUserDaoMaster;
    private static DaoSession mVipUserDaoSession;

    public static DatabaseManager getInstance() {
        return manager;
    }

    private DatabaseManager() {
        setDebug();
    }

    public DatabaseManager init(Context context) {
        this.context = context;
        return this;
    }

    public DaoMaster getVipUserDaoMaster() {
        if (mVipUserDaoMaster == null) {
            mVipUserHelper = new DbUpdateHelper(context, AppConfig.DATABASE_FILE_NAME, null);
            mVipUserDaoMaster = new DaoMaster(mVipUserHelper.getWritableDatabase());
        }
        return mVipUserDaoMaster;
    }

    public DaoSession getVipUserDaoSession() {
        if (mVipUserDaoSession == null) {
            if (mVipUserDaoMaster == null) {
                mVipUserDaoMaster = getVipUserDaoMaster();
            }
            mVipUserDaoSession = mVipUserDaoMaster.newSession();
        }
        return mVipUserDaoSession;
    }

    public void closeVipUserConnection() {
        if (mVipUserDaoSession != null) {
            mVipUserDaoSession.clear();
            mVipUserDaoSession = null;
        }
        if (mVipUserHelper != null) {
            mVipUserHelper.close();
            mVipUserHelper = null;
        }
    }

    public void setDebug() {
        if (BuildConfig.DEBUG) {
            QueryBuilder.LOG_SQL = true;
            QueryBuilder.LOG_VALUES = true;
        }
    }
}

File: app/src/main/java/com/jintao/vipmanager/database/DatabaseManager.kt
package com.jintao.vipmanager.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.jintao.vipmanager.database.bean.DbConvertGoodsInfo
import com.jintao.vipmanager.database.bean.DbUserConsumeInfo
import com.jintao.vipmanager.database.bean.DbVipUserInfo
import com.jintao.vipmanager.database.dao.DbConvertGoodsInfoDao
import com.jintao.vipmanager.database.dao.DbUserConsumeInfoDao
import com.jintao.vipmanager.database.dao.DbVipUserInfoDao
import com.jintao.vipmanager.utils.AppConfig

@Database(entities = [
    DbVipUserInfo::class,
    DbUserConsumeInfo::class,
    DbConvertGoodsInfo::class
], version = 4, exportSchema = false)
abstract class DatabaseManager : RoomDatabase() {

    companion object {

        @Volatile
        private var INSTANCE : DatabaseManager? = null

        fun getDatabase(context:Context): DatabaseManager {
            if (INSTANCE ==null) {
                synchronized(this) {
                    if (INSTANCE ==null) {
                        INSTANCE = Room.databaseBuilder(context.getApplicationContext(), DatabaseManager::class.java, AppConfig.DATABASE_FILE_NAME)
                            .addCallback(sRoomDatabaseCallback)
//                            .allowMainThreadQueries()
                            .fallbackToDestructiveMigration()
//                            .addMigrations(MIGRATION_4_5)
                            .build()
                    }
                }
            }
            return INSTANCE!!
        }

        private val sRoomDatabaseCallback = object : RoomDatabase.Callback() {
            override fun onOpen(db: SupportSQLiteDatabase) {
                super.onOpen(db)
            }
        }

        private val MIGRATION_4_5 = object :Migration(4,5) {
            override fun migrate(db: SupportSQLiteDatabase) {

            }
        }
    }

    abstract fun vipUserInfoDao():DbVipUserInfoDao
    abstract fun userConsumeInfoDao():DbUserConsumeInfoDao
    abstract fun convertGoodsInfoDao():DbConvertGoodsInfoDao

}
